File: iOS/Swift/ProductsMVP/presenter/ProductDetailPresenter.swift
import Foundation

public class ProductDetailPresenter {
    private let productDetailView: ProductDetailView
    private let product: Product?
    
    public init(productDetailView: ProductDetailView, product: Product?) {
        self.productDetailView = productDetailView
        self.product = product
    }
    
    public func onStart() {
        if let product = product {
            productDetailView.show(product: product)
        } else {
            productDetailView.showErrorWith(message: "Product not valid")
        }
    }
}

File: Android/Java/ProductMVP/app/src/main/java/it/chicio/productmvp/presenter/ProductDetailPresenter.java
package it.chicio.productmvp.presenter;

import it.chicio.productmvp.model.Product;
import it.chicio.productmvp.view.ProductDetailView;

public class ProductDetailPresenter {
    private ProductDetailView productDetailView;
    private Product product;

    public ProductDetailPresenter(ProductDetailView productDetailView, Product product) {
        this.productDetailView = productDetailView;
        this.product = product;
    }

    public void onStart() {
        productDetailView.show("Product");
        if (isValidProduct()) {
            productDetailView.show(product);
        } else {
            productDetailView.showErrorWith("Product not valid");
        }
    }

    private boolean isValidProduct() {
        return product != null;
    }
}